### ISSUE TITLE
flutter pub get fails on Flutter plugins supporting only v2 Android Embedding API

### ISSUE THREAD

**Issue by hitman-codehq** [flutter/flutter#66561]
This is a very specific bug that only occurs in certain circumstances.

You have a Flutter plugin that supports Android Embedding API v2 but not v1
You have deleted the "example" directory that is generated by the "flutter create --template=plugin" command
When executing "flutter pub get" you will always get the following output:

$ flutter pub get
Running "flutter pub get" in Embedding... 0,4s
The plugin Embedding requires your app to be migrated to the Android embedding v2. Follow the steps on
https://flutter.dev/go/android-project-migration and re-run this command.

Unfortunately, this message is incorrect and quite misleading.

I dug into the source code and it turns out that when pub get is running, it searches for the plugin implementation, which will be in a file something like:

android/src/main/kotlin/com/my_company/my_plugin/MyPlugin.kt

It loads the text for the plugin and scans it for the word "registerWith". If it finds it, it assumes that the plugin supports the Android Embedding v1 API. If not then it assumes that it only supports the v2 API.

Then it tries to look for the following file:

example/android/src/main/AndroidManifest.xml

It then checks in this file to see what the flutterEmbedding value is. If it is not "2" then it displays the message above about needing to migrate the app.

The problem is that if you have a plugin that does not contain the example directory and the AndroidManifest.xml file then you will always get the error above. But it cannot be fixed - you can't upgrade the app to the version 2 API because the app doesn't exist...

This was causing us problems in our CI system when we tried to run dartdoc to generate documentation. dartdoc executes "flutter pub get", which fails and then dartdoc bombs out.

The workaround was to put the keyword "registerWith" somewhere in the plugin's source code.

I have attached a zip file that contains an example that shows this problem. Simple unzip it, cd into the "Embedding" directory and type "dartdoc".

I would like to get involved with Flutter and volunteer to fix this bug myself. I have already extensively investigated it as you can see, so the main thing to do now is to decide the best way to implement the bug fix. Perhaps only check the AndroidManifext.xml file if the example directory exists?

[Embedding.zip](https://github.com/flutter/flutter/files/5276876/Embedding.zip)

**Comment by TahaTesser** [comment]
Hi @hitman-codehq 
I am guessing you mean `flutter pub get`, I did execute I am getting same error 
```
Running "flutter pub get" in Embedding...                           1.3s
━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
Warning
──────────────────────────────────────────────────────────────────────────────
Your Flutter application is created using an older version of the Android
embedding. It's being deprecated in favor of Android embedding v2. Follow the
steps at

https://flutter.dev/go/android-project-migration

to migrate your project.
━━━━━━━━━━━━━━━━━━━━━━━━━━━
```

<details> 
<summary>flutter doctor -v</summary> 

```bash
[✓] Flutter (Channel master, 1.22.0-10.0.pre.341, on Mac OS X 10.15.6 19G2021
    x86_64, locale en-GB)
    • Flutter version 1.22.0-10.0.pre.341 at
      /Users/tahatesser/Code/flutter_master
    • Framework revision 67f51e8ec2 (82 minutes ago), 2020-09-24 10:02:03 -0400
    • Engine revision dd35b5b659
    • Dart version 2.10.0 (build 2.10.0-156.0.dev)

[✓] Android toolchain - develop for Android devices (Android SDK version 30.0.2)
    • Android SDK at /Users/tahatesser/Code/sdk
    • Platform android-30, build-tools 30.0.2
    • ANDROID_HOME = /Users/tahatesser/Code/sdk
    • Java binary at: /Applications/Android
      Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build
      1.8.0_242-release-1644-b3-6222593)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 12.0)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 12.0, Build version 12A7209
    • CocoaPods version 1.9.3

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 4.0)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin version 49.0.2
    • Dart plugin version 193.7547
    • Java version OpenJDK Runtime Environment (build
      1.8.0_242-release-1644-b3-6222593)

[✓] VS Code (version 1.49.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.14.1

 
[✓] Connected device (4 available)            
    • RMX2001 (mobile) • EUYTFEUSQSRGDA6D • android-arm64  • Android 10 (API 29)
    • macOS (desktop)  • macos            • darwin-x64     • Mac OS X 10.15.6
      19G2021 x86_64
    • Web Server (web) • web-server       • web-javascript • Flutter Tools
    • Chrome (web)     • chrome           • web-javascript • Google Chrome
      85.0.4183.121

• No issues found!
```

</details> 

Can you please provide your `flutter doctor -v` 
Thank you

**Comment by hitman-codehq** [comment]
$ flutter doctor -v
[✓] Flutter (Channel stable, 1.20.4, on Mac OS X 10.15.6 19G2021, locale de-DE)
    • Flutter version 1.20.4 at /Users/cward/flutter
    • Framework revision fba99f6cf9 (vor 10 Tagen), 2020-09-14 15:32:52 -0700
    • Engine revision d1bc06f032
    • Dart version 2.9.2

 
[✓] Android toolchain - develop for Android devices (Android SDK version 29.0.3)
    • Android SDK at /Users/cward/Library/Android/sdk
    • Platform android-29, build-tools 29.0.3
    • ANDROID_HOME = /Users/cward/Library/Android/sdk
    • ANDROID_SDK_ROOT = /Users/cward/Library/Android/sdk
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6222593)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 12.0)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 12.0, Build version 12A7209
    • CocoaPods version 1.9.1

[✓] Android Studio (version 4.0)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin version 49.0.2
    • Dart plugin version 193.7547
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6222593)

[✓] VS Code (version 1.49.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.14.1

[✓] Connected device (1 available)
    • SM G960F (mobile) • 1c614f653c047ece • android-arm64 • Android 10 (API 29)

**Comment by Mohit020888** [comment]
Even I am running into the above issue when trying to run pub get in my flutter project.
![Android_Embedding_Error](https://user-images.githubusercontent.com/63721227/94363523-5e516900-00e0-11eb-97e3-b6131b0b678d.PNG)

**Comment by jmagman** [comment]
> I would like to get involved with Flutter and volunteer to fix this bug myself. I have already extensively investigated it as you can see, so the main thing to do now is to decide the best way to implement the bug fix. Perhaps only check the AndroidManifext.xml file if the example directory exists?

You certainly have extensively investigated!  We always welcome pull requests.  Check out https://github.com/flutter/flutter/blob/master/CONTRIBUTING.md and https://github.com/flutter/flutter/wiki/The-flutter-tool.

Please also consider writing a unit test similar to https://github.com/flutter/flutter/blob/bcd0959ac3dfe7740e0d9ba246d195fd04039977/packages/flutter_tools/test/general.shard/plugins_test.dart#L634 to test the new behavior.

**Comment by matthew-carroll** [comment]
@xster @blasten I just ran into this issue.

Repro steps:
 1. Create a new Flutter plugin
 2. `flutter create --template=plugin --platforms=android .`
 3. Delete the `/example` directory
 4. Run `flutter pub get`

Temporary workaround:
Put an /example app project into the plugin directory.

BTW - it looks like the concept of `/example` is hard-coded into the Flutter tool. Do you know the rationale behind that? I'm not sure that we should hard-code a singular acceptable name for an example project. Could we add that as a configuration to the `pubspec` instead?

**Comment by xster** [comment]
Trying to clarify what the request is. 

1- it sounds like you guys are saying pub get fails. Though I'm not clear on what the failure message is. Is it the v1 usage warning seen in the screenshots above? Are those blocking failures or just warnings?
2- Are those warning themselves incorrect as well? i.e. you're actually using v2 embeddings in the plugins but once you delete the example, it incorrectly thinks you're using v1 embeddings?

**Comment by pedromassangocode** [comment]
Without additional information, we are unfortunately not sure how to resolve this issue. 
We are therefore reluctantly going to close this bug for now. 
Please don't hesitate to comment on the bug if you have any more information for us; we will reopen it right away!  
Thanks for your contribution.

Could everyone who still has this problem please file a new issue with the exact description of what happens, logs, and the output of `flutter doctor -v`.
All system setups can be slightly different, so it's always better to open new issues and reference related issues.

**Comment by mrifni** [comment]
> Without additional information, we are unfortunately not sure how to resolve this issue.
> We are therefore reluctantly going to close this bug for now.
> Please don't hesitate to comment on the bug if you have any more information for us; we will reopen it right away!
> Thanks for your contribution.
> 
> Could everyone who still has this problem please file a new issue with the exact description of what happens, logs, and the output of `flutter doctor -v`.
> All system setups can be slightly different, so it's always better to open new issues and reference related issues.

this bug is still happening in the latest stable build and very easy to reproduce, why you guys closing the bug ?

**Comment by jmagman** [comment]
@mrifni This bug was closed because the Flutter team was waiting on the response to this question:

> Trying to clarify what the request is.
> 
> 1- it sounds like you guys are saying pub get fails. Though I'm not clear on what the failure message is. Is it the v1 usage warning seen in the screenshots above? Are those blocking failures or just warnings?
> 2- Are those warning themselves incorrect as well? i.e. you're actually using v2 embeddings in the plugins but once you delete the example, it incorrectly thinks you're using v1 embeddings?

**Comment by mrifni** [comment]
it shouldn't get closed if it's possible to reproduce @jmagman

**Comment by pedromassangocode** [comment]
> Trying to clarify what the request is.
> 
> 1- it sounds like you guys are saying pub get fails. Though I'm not clear on what the failure message is. Is it the v1 usage warning seen in the screenshots above? Are those blocking failures or just warnings?
> 2- Are those warning themselves incorrect as well? i.e. you're actually using v2 embeddings in the plugins but once you delete the example, it incorrectly thinks you're using v1 embeddings?

@mrifni  you can help by answering those questions if you are facing this same issue.

**Comment by mrifni** [comment]
@pedromassangocode answers based on my try

1- it sounds like you guys are saying pub get fails. Though I'm not clear on what the failure message is. Is it the v1 usage warning seen in the screenshots above? Are those blocking failures or just warnings?
`blocking failures`

2- Are those warning themselves incorrect as well? i.e. you're actually using v2 embeddings in the plugins but once you delete the example, it incorrectly thinks you're using v1 embeddings?
`you're actually using v2 embeddings in the plugins but once you delete the example, it incorrectly thinks you're using v1 embeddings?` YES and YES

**Comment by mrifni** [comment]
as I said again its pretty easy to reproduce this error if @pedromassangocode want to really see.

01. flutter create --template=plugin
02. Run the existing platform version check unit test and it works.
03. Delete the example folder and run the existing platform version check unit test and it fails.
04. Put an empty example folder and run the same unit test again it passes.

pretty simple steps to follow.

thanks

**Comment by hitman-codehq** [comment]
I am quite surprised that this was closed as I thought I was very clear in my initial (very long) description and supplied example.

The problem, as Mohamed says, is that Flutter assumes that you have an example directory in your plugin and if you don't then pub get fails and a completely bogus message is printed.  The message is misleading as it says to upgrade an application that doesn't exist to v2 (i.e. what is usually in the example directory) and the pub get failing causes your build to fail, which is a particular problem if you are building in Jenkins.

So the solution would be simply to check for the existence of the example directory and skip the check if it doesn't exist.

**Comment by jmagman** [comment]
> I am quite surprised that this was closed as I thought I was very clear in my initial (very long) description and supplied example.

You were clear, and it shouldn't have been closed.  More information was requested and the `waiting for customer response` label was applied, which indicates the bug should be closed if the author doesn't update in 21 days.  In this case that label was overkill since your description included a retailed reproduction case.  Sorry about that.

**Comment by freekash** [comment]
i am also facing same problem , what i the soluion?

**Comment by orestesgaolin** [comment]
> i am also facing same problem , what i the soluion?

Adding `example` for the plugin if previously removed. You can do it by going to the plugin directory and recreating plugin files:

```
flutter create .
```

Make sure to save/commit your changes before as it will generate a lot of files.

**Comment by NdegwaJulius** [comment]
I'm getting this error when trying to use AudioCache in my project.
The plugin `audioplayers` requires your app to be migrated to the Android embedding v2
